### ISSUE TITLE
Support multiple directory trees and filename changes

### ISSUE THREAD

**Issue by oconnore** [tschaub/gulp-newer#1]
gulp-newer currently only supports single directory targets with direct file matches, and single file 'all' targets. For my build I need to do things like:

```
./public/**/*.html --> ./working/
./public/**/*.styl --> ./working/**/*.css
 --- results in ---
./public/a/b/c/index.html -> ./working/a/b/c/index.html
./public/a/b/c/index.html -> ./working/a/b/c/index.html
./public/css/stylus/main.styl -> ./working/css/stylus/main.css
```

I setup my build without using gulp-newer, but would like to contribute back to make this module work better for everyone (also me) in the future.

Currently, my API for these things looks like:
```javascript
.pipe(newer(function(srcPath) {
  return destPath;
}))
```

That's really [infinitely] flexible, but maybe preserving a nicer API is a good idea as well. Lets start with:

#### Nested directories

If we are OK requiring that the directory structure in the target is built beforehand, we can use it to find the deepest possible match (from the right) relative to the base directory, and require no change to the gulp-newer API. This heuristic should work most of the time:
```
directories:
./working/a/b/c
./working/a/b/d
./working/a/b/c/e
./public/a/b/c/index.html ->./working/a/b/c/index.html
```
(Until joyent/node#6662 is resolved, this may require memoizing stat calls -- I had to do that for my build, but I would want to do more profiling before I committed that assumption to a public project.)

#### Pathname changes

Changes of file name (extension, append, replace, etc.) is a bit more complicated, and while I can imagine some ways to integrate it with the existing single-string configuration style, that seems messy, non obvious, and prone to error. For example, attempts to correlate output files on the first build with source files, and attempts to parse a change in filename from a globbed output pattern both seem really unpleasant to implement, support, and use. 

We also (I think?) would like to not require users to write a conversion function when possible, but should allow it when necessary. Here are some thoughts:

- Allow an ```options``` object to be passed as ```newer(destBase, { /* options */ })```
- Allow rewriting the path and the filename separately. Accept a ```sourceBase``` option to change the cwd. 
```javascript
options = {
  destBase: /* allow the options object to be passed by itself */,
  srcBase: /* base directory of source files */,
  path: /* change path from relative to sourceBase */,
  filename: /* change filename */,
  fullpath: /* change path and filename at the same time */
}
```
- for ```path```, ```filename```, and ```fullpath``` mutators:
   - accept handlebars style string ```{filebase}.css``` with keys like:
      - ```{path [0-9]+(-[0-9]+)?}```
      - ```{filebase}``` and ```{fileext}```
   - accept an ```s/from/to/``` style regex
   - accept a function that converts from source -> dest paths

#### Discussion

What do you think? Would you accept a pull request for something like this + feedback?

**Comment by agborkowski** [comment]
@oconnore TC !
``````js
gulp.task('images', function() {
   return gulp.src('resources/uploads/**/*')
    //.pipe(changed('webroot/galleries'))
    //.pipe(cache('images'))
    .pipe(newer({
      dest: 'webroot/galleries',
      map: function(pathDir) { 
        pathDir = 
          path.dirname(pathDir) + '/50-x-' +
          path.basename(pathDir, path.extname(pathDir)) + '-thumb' + 
          path.extname(pathDir);
        return pathDir; 
      }
    }))
    .pipe(imagemin({ optimizationLevel: 3, progressive: true, interlaced: true }))
    .pipe(
      responsive(
      {
        '**/*' : [{
          width: 50,
          rename: {
            //path.dirname += "";
            prefix: "50-x-",
            suffix: "-thumb"
            //path.extname = ".md"
          }
        }]
      })
    )
    .pipe(gulp.dest('webroot/galleries'))
    .pipe(notify({ message: 'Images task complete' }));
});
`````````````

**Comment by Saturn2888** [comment]
I have one other suggestion for this functionality.

If someone deletes the `.map` file, when I go to build, any file that hadn't had changes, such as the one w/ the deleted `.map` file wouldn't be built because I'm doing this:

```js
.pipe($.newer({
    dest: dest,
    ext: '.js'
}))
```

What I want is:
```js
ext: [
    '.js',
    '.min.js'
    '.map'
]
```

I can't figure out how to pipe in two gulp-newer operations because the first would cancel out the second if it exists. So this might be possible that way as well.

**Comment by tschaub** [comment]
@Saturn2888 - could you open a separate issue with your suggestion?  This ticket has migrated pretty far from the original title/description.  I'm going to close this as addressed by #2, #3, and #12.
